src/gradualizer_type.erl

```diff
@@ -0,0 +1,355 @@
+        {'fun', anno(), {'function', function_name(), arity()}}.
+
+-type af_remote_fun() ::
+        {'fun', anno(), {'function', module(), function_name(), arity()}}
+      | {'fun', anno(), {'function', af_atom(), af_atom(), af_integer()}}.
+
+-type af_fun() :: {'fun', anno(), {'clauses', af_clause_seq()}}.
+
+-type af_named_fun() :: {'named_fun', anno(), fun_name(), af_clause_seq()}.
+
+-type fun_name() :: atom().
+
+-type af_clause() ::
+        {'clause', anno(), [af_pattern()], af_guard_seq(), af_body()}.
+
+-type af_body() :: [abstract_expr(), ...].
+
+-type af_guard_seq() :: [af_guard()].
+
+-type af_guard() :: [af_guard_test(), ...].
+
+-type af_guard_test() :: af_literal()
+                       | af_variable()
+                       | af_tuple(af_guard_test())
+                       | af_nil()
+                       | af_cons(af_guard_test())
+                       | af_bin(af_guard_test())
+                       | af_binary_op(af_guard_test())
+                       | af_unary_op(af_guard_test())
+                       | af_record_creation(af_guard_test())
+                       | af_record_index()
+                       | af_record_field_access(af_guard_test())
+                       | af_map_creation(abstract_expr())
+                       | af_map_update(abstract_expr())
+                       | af_guard_call()
+                       | af_remote_guard_call().
+
+-type af_record_field_access(T) ::
+        {'record_field', anno(), T, record_name(), af_field_name()}.
+
+-type af_map_creation(T) :: {'map', anno(), [af_assoc(T)]}.
+
+-type af_map_update(T) :: {'map', anno(), T, [af_assoc(T)]}.
+
+-type af_assoc(T) :: {'map_field_assoc', anno(), T, T}
+                   | af_assoc_exact(T).
+
+-type af_assoc_exact(T) :: {'map_field_exact', anno(), T, T}.
+
+-type af_guard_call() :: {'call', anno(), function_name(), [af_guard_test()]}.
+
+-type af_remote_guard_call() ::
+        {'call', anno(),
+         {'remote', anno(), af_lit_atom('erlang'), af_atom()},
+         [af_guard_test()]}.
+
+-type af_pattern() :: af_literal()
+                    | af_match(af_pattern())
+                    | af_variable()
+                    | af_tuple(af_pattern())
+                    | af_nil()
+                    | af_cons(af_pattern())
+                    | af_bin(af_pattern())
+                    | af_binary_op(af_pattern())
+                    | af_unary_op(af_pattern())
+                    | af_record_creation(af_pattern())
+                    | af_record_index()
+                    | af_map_pattern().
+
+-type af_record_index() ::
+        {'record_index', anno(), record_name(), af_field_name()}.
+
+-type af_record_creation(T) ::
+        {'record', anno(), record_name(), [af_record_field(T)]}.
+
+-type af_record_field(T) :: {'record_field', anno(), af_field_name(), T}.
+
+-type af_map_pattern() ::
+        {'map', anno(), [af_assoc_exact(abstract_expr())]}.
+
+-type abstract_type() :: af_annotated_type()
```

Comment:
I think we also want to include ` af_constrained_function_type()` in our `abstract_type()` (this can go in a separate commit as it is a modification of the upstream version)

we can possibly have two `type()`s (not sure about the naming)
- one which can come from an AST
- one extended that Gradualizer can handle internally (which have intersection, bounded_fun types, maybe cons type and any other additions)

Reply:
@gomoripeti can you expand a bit about what different situations you imagine we should use the two different `type()` types?

Reply:
I imagined that certain functions could avoid handling the extended parts like intersection. Those for example which directly receive a type from a `-type` definition from source. Because you cannot express intersection in general there (yet)

But I realise we cannot define it as `extended_type() :: type() | <extra stuff>` as they are recursive, so would need to duplicate all the subdefinitions of `abstract_type()` which is not desirable. Probably better to handle the extra additions (like intersection) everywhere.

So please ignore this idea.

ps: if we would change the internal representation (eg to records) we would know which functions should still handle the AST type and which the internal type. Sometimes it is useful to encode such additional information in the type definitions even if the actual types don't differ. (just like `temperature_C() :: float()` and `temperature_F() :: float()`)